core/src/prediction/LocalLinearPredictionStrategy.cpp

```diff
@@ -46,35 +46,35 @@ LocalLinearPredictionStrategy::LocalLinearPredictionStrategy(const Data* origina
 std::vector<double> LocalLinearPredictionStrategy::predict(size_t sampleID,
                                                            const std::unordered_map<size_t, double>& weights_by_sampleID,
                                                            const Observations& observations) {
-  size_t n = observations.get_num_samples();
   size_t num_variables = linear_correction_variables.size();
-
-  Eigen::MatrixXd weights = Eigen::MatrixXd::Zero(n,n);
-
-  for (auto it = weights_by_sampleID.begin(); it != weights_by_sampleID.end(); ++it){
-    size_t i = it->first;
-    double weight = it->second;
-    weights(i,i) = weight;
+  size_t num_nonzero_weights = weights_by_sampleID.size();
+
+  std::vector<size_t> indices(num_nonzero_weights);
+  Eigen::MatrixXd weights_vec = Eigen::VectorXd::Zero(num_nonzero_weights);
+  size_t weights_index = 0;
```

Comment:
(nitpick) it would be more consistent with the loops below if this were named `i`, and we did `size_t index = it.first;`.